### PR TITLE
Make TestLauncher test debugging compatible with configuration cache

### PR DESCRIPTION
**Pull request by donat**
The debug configuration was ignored when the configuration cache entry was reused.

Partially fixes https://github.com/gradle/gradle/issues/26366

### CODE REVIEW

**Comment by donat** [comment]
@bot-gradle test RFN

**Review by hegyibalint — APPROVED** [review]

**Comment by donat** [comment]
@bot-gradle test RFN

**Comment by donat** [comment]
@bot-gradle test and merge

**Comment by ljacomet** [comment]
@bot-gradle test and merge

**Comment by donat** [comment]
@bot-gradle cancel

**Comment by donat** [comment]
@bot-gradle test RFN

File: subprojects/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r85/TestLauncherDebugCrossVersionTest.groovy [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+    private String runTaskAndTestClassUsing(boolean debugMode) {
+        def stdout = new ByteArrayOutputStream()
+        withConnection { ProjectConnection connection ->
+            def tee = new TeeOutputStream(System.out, stdout)
+            def testLauncher = connection.newTestLauncher()
+                .withTaskAndTestClasses(':app:test', Arrays.asList('TestClass1'))
+                .setStandardOutput(tee)
+            if (debugMode) {
+                // JDWPUtil is flaky https://github.com/gradle/gradle/issues/26366
+                testLauncher.debugTestsOn(49052 + new Random(System.currentTimeMillis()).nextInt(100))
```
Comment (tresat) [inline subprojects/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r85/TestLauncherDebugCrossVersionTest.groovy:66]: ❗ This makes the test non-deterministic, harder to debug and more likely to become flaky.  Is a static seed possible, perhaps incrementing until a free port is found?

Comment (donat) [reply subprojects/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r85/TestLauncherDebugCrossVersionTest.groovy:66]: This is a workaround until https://github.com/gradle/gradle-private/issues/3986 is fixed (need to adjust the URL there). There's something funky about detecting free ports on CI, thence the workaround with a random number.

Comment (donat) [reply subprojects/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r85/TestLauncherDebugCrossVersionTest.groovy:66]: Btw this test doesn't allocate any port at all, just needs the debugger to not actively listen on the selected random port. Any other process using that port is fine.

File: subprojects/tooling-api-builders/src/main/java/org/gradle/tooling/internal/provider/runner/TestExecutionBuildConfigurationAction.java [hunk 2]
```diff
@@ -109,6 +114,7 @@ private void configureTestTasksForTestDescriptors(Context context) {
         for (final Map.Entry<String, List<InternalJvmTestRequest>> entry : taskAndTests.entrySet()) {
             String testTaskPath = entry.getKey();
             for (AbstractTestTask testTask : queryTestTasks(context, testTaskPath)) {
+                configureTestTask(testTask);
```
Comment (mlopatkin) [inline subprojects/tooling-api-builders/src/main/java/org/gradle/tooling/internal/provider/runner/TestExecutionBuildConfigurationAction.java:117]: :thinking: I wonder if we need to run `configureTestTasks` in `applyTasksTo` now, we're re-running configuration in `postProcessExecutionPlan` anyway.

Comment (donat) [reply subprojects/tooling-api-builders/src/main/java/org/gradle/tooling/internal/provider/runner/TestExecutionBuildConfigurationAction.java:117]: That's a good point, I'll check that.

File: subprojects/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r85/TestLauncherDebugCrossVersionTest.groovy (hunk 1 above)
Comment (donat) [reply subprojects/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r85/TestLauncherDebugCrossVersionTest.groovy]: Fixed

File: subprojects/tooling-api-builders/src/main/java/org/gradle/tooling/internal/provider/runner/TestExecutionBuildConfigurationAction.java (hunk 2 above)
Comment (donat) [reply subprojects/tooling-api-builders/src/main/java/org/gradle/tooling/internal/provider/runner/TestExecutionBuildConfigurationAction.java:108]: You are right, there's no need to call `configureTestTask()` within `applyTasksTo()`.

**Comment by donat** [comment]
@bot-gradle test RFN

File: subprojects/tooling-api-builders/src/main/java/org/gradle/tooling/internal/provider/runner/TestExecutionBuildConfigurationAction.java (hunk 2 above)
Comment (donat) [reply subprojects/tooling-api-builders/src/main/java/org/gradle/tooling/internal/provider/runner/TestExecutionBuildConfigurationAction.java:108]: There are fest failures indicating that we cannot simply remove configureTestTask() from here.

**Comment by donat** [comment]
@bot-gradle test RFN

**Comment by donat** [comment]
@bot-gradle test RFN

**Comment by donat** [comment]
@bot-gradle test RFN

**Comment by donat** [comment]
@bot-gradle test and merge

**Pull request merged**